File: skills/docforge/scripts/provenance_frontmatter.py
#!/usr/bin/env python3
"""Restricted YAML frontmatter codec for Docforge provenance 2.0.

Emits a deterministic YAML subset (2-space indent, fixed key order, every
scalar double-quoted). Parses that subset plus plain unquoted scalars.
Rejects anchors, aliases, block scalars, multi-document markers, and
non-empty flow collections. Standard library only.
"""

from __future__ import annotations

import hashlib
import json
import re
from typing import Any

SCHEMA_VERSION = "2.0"
LEGACY_SCHEMA = "1.0"
GENERATOR_NAME = "docforge"
GENERATOR_VERSION = "2.1.0"
PROVENANCE_FIELDS = {
    "schema", "doc_id", "path", "generated_at", "generator", "tier",
    "target_depth", "graph", "sections",
}
OPTIONAL_FIELDS = {"git_commit", "content_hash", "review"}
SOURCE_ROLES = {"code", "config", "manifest", "doc", "test", "history"}
FLOW_VALUES = {"native", "derived", "none"}
BLOB = re.compile(r"^[0-9a-f]{40}$")
SCAFFOLD_TOKEN = re.compile(r"^<[A-Z][A-Z0-9_]{2,}>$")
CONTENT_HASH = re.compile(r"^sha256:[0-9a-f]{64}$")
FRONTMATTER = re.compile(r"\A---\n(.*?)\n---\n", re.DOTALL)

PROVENANCE_KEY_ORDER = [
    "schema", "doc_id", "path", "generated_at", "generator",
    "tier", "target_depth", "git_commit", "content_hash", "review",
    "graph", "sections",
]
GENERATOR_KEY_ORDER = ["name", "version"]
GRAPH_KEY_ORDER = ["provider", "flow"]
SECTION_KEY_ORDER = ["id", "sources", "unresolved"]
SOURCE_KEY_ORDER = ["path", "git_blob", "role"]
REVIEW_KEY_ORDER = ["mode", "verdict", "report"]


class YamlCodecError(ValueError):
    """Raised when restricted YAML cannot be parsed or emitted."""


def quote_scalar(value: str) -> str:
    return json.dumps(value, ensure_ascii=False)


def content_hash(body: str) -> str:
    digest = hashlib.sha256(body.encode("utf-8")).hexdigest()
    return f"sha256:{digest}"


def scaffold_provenance(
    doc_id: str,
    path: str,
    *,
    tier: str = "<TIER>",
    target_depth: str = "<TARGET_DEPTH>",
    provider: str = "<GRAPH_PROVIDER>",
    flow: str = "<FLOW_CAPABILITY>",
    generated_at: str = "<GENERATED_AT>",
) -> dict:
    return {
        "schema": SCHEMA_VERSION,
        "doc_id": doc_id,
        "path": path,
        "generated_at": generated_at,
        "generator": {"name": GENERATOR_NAME, "version": GENERATOR_VERSION},
        "tier": tier,
        "target_depth": target_depth,
        "graph": {"provider": provider, "flow": flow},
        "sections": [],
    }


def migrate_v1_to_v2(provenance: dict, body: str = "") -> dict:
    if not isinstance(provenance, dict):
        raise YamlCodecError("provenance must be an object")
    schema = provenance.get("schema")
    if schema == SCHEMA_VERSION and "generator" in provenance:
        return dict(provenance)
    if schema != LEGACY_SCHEMA and "tool_version" not in provenance and schema is not None:
        raise YamlCodecError(f"unsupported provenance schema: {schema}")
    tool_version = provenance.get("tool_version") or GENERATOR_VERSION
    if not isinstance(tool_version, str) or not tool_version:
        tool_version = GENERATOR_VERSION
    migrated = {
        "schema": SCHEMA_VERSION,
        "doc_id": provenance.get("doc_id", ""),
        "path": provenance.get("path", ""),
        "generated_at": provenance.get("generated_at", ""),
        "generator": {"name": GENERATOR_NAME, "version": tool_version},
        "tier": provenance.get("tier", ""),
        "target_depth": provenance.get("target_depth", ""),
        "graph": provenance.get("graph") if isinstance(provenance.get("graph"), dict) else {
            "provider": "<GRAPH_PROVIDER>",
            "flow": "<FLOW_CAPABILITY>",
        },
        "sections": provenance.get("sections") if isinstance(provenance.get("sections"), list) else [],
    }
    if isinstance(provenance.get("git_commit"), str):
        migrated["git_commit"] = provenance["git_commit"]
    if isinstance(provenance.get("content_hash"), str):
        migrated["content_hash"] = provenance["content_hash"]
    elif body:
        migrated["content_hash"] = content_hash(body)
    if isinstance(provenance.get("review"), dict):
        migrated["review"] = provenance["review"]
    return migrated


def _emit_mapping(lines: list[str], mapping: dict, key_order: list[str], indent: int) -> None:
    prefix = " " * indent
    for key in key_order:
        if key not in mapping:
            continue
        value = mapping[key]
        if isinstance(value, dict):
            lines.append(f"{prefix}{key}:")
            if key == "generator":
                _emit_mapping(lines, value, GENERATOR_KEY_ORDER, indent + 2)
            elif key == "graph":
                _emit_mapping(lines, value, GRAPH_KEY_ORDER, indent + 2)
            elif key == "review":
                _emit_mapping(lines, value, REVIEW_KEY_ORDER, indent + 2)
            else:
                nested_order = list(value.keys())
                _emit_mapping(lines, value, nested_order, indent + 2)
        elif isinstance(value, list):
            if not value:
                lines.append(f"{prefix}{key}: []")
                continue
            lines.append(f"{prefix}{key}:")
            for item in value:
                if isinstance(item, dict):
                    first = True
                    item_order = (
                        SECTION_KEY_ORDER if "sources" in item
                        else SOURCE_KEY_ORDER if "git_blob" in item
                        else list(item.keys())
                    )
                    for item_key in item_order:
                        if item_key not in item:
                            continue
                        item_value = item[item_key]
                        bullet = "- " if first else "  "
                        first = False
                        if isinstance(item_value, list):
                            if not item_value:
                                lines.append(f"{prefix}  {bullet}{item_key}: []")
                            else:
                                lines.append(f"{prefix}  {bullet}{item_key}:")
                                for nested in item_value:
                                    if isinstance(nested, dict):
                                        nested_first = True
                                        for nested_key in SOURCE_KEY_ORDER:
                                            if nested_key not in nested:
                                                continue
                                            nested_bullet = "- " if nested_first else "  "
                                            nested_first = False
                                            lines.append(
                                                f"{prefix}      {nested_bullet}{nested_key}: "
                                                f"{quote_scalar(str(nested[nested_key]))}"
                                            )
                                    else:
                                        lines.append(
                                            f"{prefix}      - {quote_scalar(str(nested))}"
                                        )
                        else:
                            lines.append(
                                f"{prefix}  {bullet}{item_key}: {quote_scalar(str(item_value))}"
                            )
                else:
                    lines.append(f"{prefix}  - {quote_scalar(str(item))}")
        else:
            lines.append(f"{prefix}{key}: {quote_scalar(str(value))}")


def emit_yaml(provenance: dict) -> str:
    if not isinstance(provenance, dict):
        raise YamlCodecError("provenance must be an object")
    lines = ["---", "docforge_provenance:"]
    _emit_mapping(lines, provenance, PROVENANCE_KEY_ORDER, 2)
    lines.append("---")
    lines.append("")
    return "\n".join(lines)


def wrap_document(provenance: dict, body: str) -> str:
    if body.startswith("\n"):
        return emit_yaml(provenance) + body.lstrip("\n")
    if body and not body.endswith("\n"):
        body = body + "\n"
    return emit_yaml(provenance) + body


def _reject_unsupported(raw: str) -> None:
    for line in raw.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        if stripped == "---" or stripped.startswith("..."):
            raise YamlCodecError("unsupported yaml construct: multi-document marker")
        if "&" in stripped or "*" in stripped:
            raise YamlCodecError("unsupported yaml construct: anchor or alias")
        if stripped.endswith("|") or stripped.endswith(">") or stripped.endswith("|-") or stripped.endswith(">-"):
            raise YamlCodecError("unsupported yaml construct: block scalar")
        if re.search(r":\s*[\[{].+[\]}]", stripped):
            raise YamlCodecError("unsupported yaml construct: non-empty flow collection")


def _parse_scalar(raw: str) -> Any:
    value = raw.strip()
    if value == "[]":
        return []
    if value == "{}":
        return {}
    if value in {"null", "~"}:
        return None
    if value in {"true", "false"}:
        return value == "true"
    if len(value) >= 2 and value[0] == value[-1] and value[0] in {'"', "'"}:
        if value[0] == '"':
            return json.loads(value)
        return value[1:-1]
    return value


def _indent_of(line: str) -> int:
    return len(line) - len(line.lstrip(" "))


def parse_yaml_mapping(raw: str) -> dict:
    _reject_unsupported(raw)
    lines = [line.rstrip("\n") for line in raw.splitlines() if line.strip() and not line.strip().startswith("#")]
    root: dict[str, Any] = {}
    stack: list[tuple[int, Any]] = [(-1, root)]
    i = 0
    while i < len(lines):
        line = lines[i]
        indent = _indent_of(line)
        content = line.strip()
        while len(stack) > 1 and indent <= stack[-1][0]:
            stack.pop()
        parent = stack[-1][1]
        if content.startswith("- "):
            if not isinstance(parent, list):
                raise YamlCodecError("list item without list parent")
            item_content = content[2:]
            if ": " in item_content or item_content.endswith(":"):
                item: dict[str, Any] = {}
                parent.append(item)
                stack.append((indent, item))
                if item_content.endswith(":"):
                    key = item_content[:-1].strip()
                    # Look ahead for nested value or empty list.
                    if i + 1 < len(lines) and _indent_of(lines[i + 1]) > indent:
                        nxt = lines[i + 1].strip()
                        if nxt.startswith("- "):
                            item[key] = []
                            stack.append((indent + 2, item[key]))
                        elif nxt == "[]":
                            i += 1
                            item[key] = []
                        else:
                            item[key] = {}
                            stack.append((indent + 2, item[key]))
                    else:
                        item[key] = {}
                else:
                    key, raw_value = item_content.split(":", 1)
                    item[key.strip()] = _parse_scalar(raw_value)
            else:
                parent.append(_parse_scalar(item_content))
            i += 1
            continue
        if ": " in content or content.endswith(":"):
            if not isinstance(parent, dict):
                raise YamlCodecError("mapping entry without mapping parent")
            if content.endswith(":"):
                key = content[:-1].strip()
                if i + 1 < len(lines) and _indent_of(lines[i + 1]) > indent:
                    nxt = lines[i + 1].strip()
                    if nxt == "[]":
                        i += 1
                        parent[key] = []
                    elif nxt.startswith("- "):
                        parent[key] = []
                        stack.append((indent, parent[key]))
                    else:
                        parent[key] = {}
                        stack.append((indent, parent[key]))
                else:
                    parent[key] = {}
            else:
                key, raw_value = content.split(":", 1)
                parent[key.strip()] = _parse_scalar(raw_value)
            i += 1
            continue
        raise YamlCodecError(f"unsupported yaml construct: {content}")
    return root


def split_frontmatter(text: str) -> tuple[str | None, str, int]:
    match = FRONTMATTER.match(text)
    if not match:
        return None, text, 0
    return match.group(1), text[match.end():], match.end()


def parse_frontmatter(text: str) -> tuple[str, dict | None, int]:
    """Return (state, provenance_or_none, body_start).

    States: missing, unparseable, legacy, obsolete, ok.
    """
    raw, _body, end = split_frontmatter(text)
    if raw is None:
        return "missing", None, 0
    stripped = raw.lstrip()
    if stripped.startswith("{"):
        try:
            data = json.loads(raw)
        except json.JSONDecodeError:
            return "unparseable", None, end
        provenance = data.get("docforge_provenance") if isinstance(data, dict) else None
        if not isinstance(provenance, dict):
            return "missing", None, end
        if "schema" not in provenance:
            return "legacy", provenance, end
        if provenance.get("schema") == LEGACY_SCHEMA or "tool_version" in provenance:
            return "obsolete", provenance, end
        return "obsolete", provenance, end
    try:
        data = parse_yaml_mapping(raw)
    except YamlCodecError:
        return "unparseable", None, end
    provenance = data.get("docforge_provenance") if isinstance(data, dict) else None
    if not isinstance(provenance, dict):
        return "missing", None, end
    if "schema" not in provenance:
        return "legacy", provenance, end
    if provenance.get("schema") != SCHEMA_VERSION or "tool_version" in provenance:
        return "obsolete", provenance, end
    return "ok", provenance, end


def rewrite_frontmatter(text: str, provenance: dict) -> str:
    raw, body, _end = split_frontmatter(text)
    if raw is None:
        return wrap_document(provenance, text)
    return emit_yaml(provenance) + body

File: skills/docforge/_shared/runtime/common/js/provenance_frontmatter.js
#!/usr/bin/env node
"use strict";
/* Restricted YAML frontmatter codec for Docforge provenance 2.0.
 * Emits a deterministic YAML subset and parses that subset plus plain
 * unquoted scalars. Standard library / Node built-ins only.
 */

const crypto = require("crypto");

const SCHEMA_VERSION = "2.1";
const SUPPORTED_SCHEMA_VERSIONS = new Set(["2.0", "2.1"]);
const LEGACY_SCHEMA = "1.0";
const GENERATOR_NAME = "docforge";
const GENERATOR_VERSION = "2.8.0";
const PROVENANCE_FIELDS = new Set([
  "schema", "doc_id", "path", "generated_at", "generator", "tier",
  "target_depth", "graph", "sections",
]);
const OPTIONAL_FIELDS = new Set(["git_commit", "content_hash", "review"]);
const SOURCE_ROLES = new Set(["code", "config", "manifest", "doc", "test", "history"]);
const FLOW_VALUES = new Set(["native", "derived", "none"]);
const BLOB = /^[0-9a-f]{40}$/;
const SCAFFOLD_TOKEN = /^<[A-Z][A-Z0-9_]{2,}>$/;
const CONTENT_HASH = /^sha256:[0-9a-f]{64}$/;

const PROVENANCE_KEY_ORDER = [
  "schema", "doc_id", "path", "generated_at", "generator",
  "tier", "target_depth", "git_commit", "content_hash", "review",
  "graph", "sections",
];
const GENERATOR_KEY_ORDER = ["name", "version"];
const GRAPH_KEY_ORDER = ["provider", "flow"];
const SECTION_KEY_ORDER = ["id", "sources", "unresolved"];
const SOURCE_KEY_ORDER = [
  "path", "git_blob", "git_blob_normalized", "evidence_range", "range_blob", "role",
];
const EVIDENCE_RANGE_KEY_ORDER = ["start", "end"];
const REVIEW_KEY_ORDER = ["mode", "verdict", "report"];

class YamlCodecError extends Error {
  constructor(message) {
    super(message);
    this.name = "YamlCodecError";
  }
}

function quoteScalar(value) {
  return JSON.stringify(String(value));
}

function contentHash(body) {
  return `sha256:${crypto.createHash("sha256").update(body, "utf8").digest("hex")}`;
}

function scaffoldProvenance(docId, pathValue, options = {}) {
  return {
    schema: SCHEMA_VERSION,
    doc_id: docId,
    path: pathValue,
    generated_at: options.generated_at || "<GENERATED_AT>",
    generator: { name: GENERATOR_NAME, version: GENERATOR_VERSION },
    tier: options.tier || "<TIER>",
    target_depth: options.target_depth || "<TARGET_DEPTH>",
    graph: {
      provider: options.provider || "<GRAPH_PROVIDER>",
      flow: options.flow || "<FLOW_CAPABILITY>",
    },
    sections: [],
  };
}

function inferSourceRole(pathValue) {
  const lower = String(pathValue).replace(/\\/g, "/").toLowerCase();
  const name = lower.split("/").pop() || "";
  if (["/test/", "/tests/", "/__tests__/", "_test.", ".test.", ".spec."].some((m) => lower.includes(m))) {
    return "test";
  }
  if (
    [
      "package.json", "pyproject.toml", "cargo.toml", "go.mod", "gemfile",
      "pom.xml", "composer.json", "build.gradle", "build.gradle.kts",
    ].includes(name)
    || name.endsWith(".csproj")
    || name.endsWith(".gemspec")
  ) {
    return "manifest";
  }
  if (
    [".yml", ".yaml", ".toml", ".ini", ".cfg", ".conf", ".env", ".properties"].some((ext) => name.endsWith(ext))
    || `/${lower}/`.includes("/config/")
    || name.startsWith(".")
  ) {
    return "config";
  }
  if ([".md", ".rst", ".txt", ".adoc"].some((ext) => name.endsWith(ext)) || lower.startsWith("docs/")) {
    return "doc";
  }
  return "code";
}

function inferProviderFromSnapshot(snapshot) {
  const lower = String(snapshot).replace(/\\/g, "/").toLowerCase();
  if (`/${lower}`.includes("/.ua/") || lower.startsWith(".ua/") || lower.includes("understand-anything")) {
    return "understand-anything";
  }
  if (lower.includes("gitnexus")) return "gitnexus";
  if (lower.includes("codegraph")) return "codegraph";
  return null;
}

function normalizeSections(sections) {
  if (!Array.isArray(sections)) return [];
  const normalized = [];
  for (const section of sections) {
    if (!section || typeof section !== "object" || Array.isArray(section)) continue;
    const sources = [];
    if (Array.isArray(section.sources)) {
      for (const source of section.sources) {
        if (!source || typeof source !== "object" || Array.isArray(source)) continue;
        const pathValue = source.path;
        const blob = source.git_blob;
        if (typeof pathValue !== "string" || !pathValue || typeof blob !== "string" || !blob) continue;
        const entry = { path: pathValue, git_blob: blob };
        const normalizedBlob = source.git_blob_normalized;
        if (typeof normalizedBlob === "string" && BLOB.test(normalizedBlob)) {
          entry.git_blob_normalized = normalizedBlob;
        }
        const evidenceRange = source.evidence_range;
        const rangeBlob = source.range_blob;
        if (
          evidenceRange && typeof evidenceRange === "object" && !Array.isArray(evidenceRange)
          && typeof evidenceRange.start === "string" && /^[1-9][0-9]*$/.test(evidenceRange.start)
          && typeof evidenceRange.end === "string" && /^[1-9][0-9]*$/.test(evidenceRange.end)
          && Number(evidenceRange.end) >= Number(evidenceRange.start)
          && typeof rangeBlob === "string" && BLOB.test(rangeBlob)
        ) {
          entry.evidence_range = { start: evidenceRange.start, end: evidenceRange.end };
          entry.range_blob = rangeBlob;
        }
        const role = SOURCE_ROLES.has(source.role) ? source.role : inferSourceRole(pathValue);
        entry.role = role;
        sources.push(entry);
      }
    }
    normalized.push({
      id: String(section.id || "main"),
      sources,
      unresolved: Array.isArray(section.unresolved) ? section.unresolved.slice() : [],
    });
  }
  return normalized;
}

function migrateV1ToV2(provenance, body = "", defaults = {}) {
  if (!provenance || typeof provenance !== "object" || Array.isArray(provenance)) {
    throw new YamlCodecError("provenance must be an object");
  }
  const opts = defaults && typeof defaults === "object" && !Array.isArray(defaults) ? defaults : {};
  if (SUPPORTED_SCHEMA_VERSIONS.has(provenance.schema) && provenance.generator) {
    const migrated = { ...provenance };
    if (Array.isArray(migrated.sections)) migrated.sections = normalizeSections(migrated.sections);
    return migrated;
  }
  if (
    provenance.schema !== LEGACY_SCHEMA
    && !("tool_version" in provenance)
    && provenance.schema != null
  ) {
    throw new YamlCodecError(`unsupported provenance schema: ${provenance.schema}`);
  }
  let toolVersion = provenance.tool_version || GENERATOR_VERSION;
  if (typeof toolVersion !== "string" || !toolVersion) toolVersion = GENERATOR_VERSION;
  const pathValue = provenance.path || provenance.doc || opts.path || "";
  const docId = provenance.doc_id || opts.doc_id || "";
  let graph;
  if (provenance.graph && typeof provenance.graph === "object" && !Array.isArray(provenance.graph)) {
    graph = {
      provider: String(provenance.graph.provider || opts.provider || "<GRAPH_PROVIDER>"),
      flow: String(provenance.graph.flow || opts.flow || "<FLOW_CAPABILITY>"),
    };
  } else {
    let inferred = null;
    if (typeof provenance.graph_snapshot === "string" && provenance.graph_snapshot) {
      inferred = inferProviderFromSnapshot(provenance.graph_snapshot);
    }
    graph = {
      provider: String(inferred || opts.provider || "<GRAPH_PROVIDER>"),
      flow: String(opts.flow || "<FLOW_CAPABILITY>"),
    };
    if (inferred === "understand-anything" && graph.flow === "<FLOW_CAPABILITY>") {
      graph.flow = "native";
    }
  }
  const migrated = {
    schema: SCHEMA_VERSION,
    doc_id: String(docId),
    path: String(pathValue),
    generated_at: String(provenance.generated_at || opts.generated_at || ""),
    generator: { name: GENERATOR_NAME, version: toolVersion },
    tier: String(provenance.tier || opts.tier || ""),
    target_depth: String(provenance.target_depth || opts.target_depth || ""),
    graph,
    sections: normalizeSections(provenance.sections),
  };
  if (typeof provenance.git_commit === "string") migrated.git_commit = provenance.git_commit;
  if (typeof provenance.content_hash === "string") migrated.content_hash = provenance.content_hash;
  else if (body) migrated.content_hash = contentHash(body);
  if (provenance.review && typeof provenance.review === "object") migrated.review = provenance.review;
  return migrated;
}

function emitMapping(lines, mapping, keyOrder, indent) {
  const prefix = " ".repeat(indent);
  for (const key of keyOrder) {
    if (!(key in mapping)) continue;
    const value = mapping[key];
    if (value && typeof value === "object" && !Array.isArray(value)) {
      lines.push(`${prefix}${key}:`);
      if (key === "generator") emitMapping(lines, value, GENERATOR_KEY_ORDER, indent + 2);
      else if (key === "graph") emitMapping(lines, value, GRAPH_KEY_ORDER, indent + 2);
      else if (key === "review") emitMapping(lines, value, REVIEW_KEY_ORDER, indent + 2);
      else emitMapping(lines, value, Object.keys(value), indent + 2);
    } else if (Array.isArray(value)) {
      if (!value.length) {
        lines.push(`${prefix}${key}: []`);
        continue;
      }
      lines.push(`${prefix}${key}:`);
      for (const item of value) {
        if (item && typeof item === "object" && !Array.isArray(item)) {
          let first = true;
          const itemOrder = ("sources" in item)
            ? SECTION_KEY_ORDER
            : ("git_blob" in item)
              ? SOURCE_KEY_ORDER
              : Object.keys(item);
          for (const itemKey of itemOrder) {
            if (!(itemKey in item)) continue;
            const itemValue = item[itemKey];
            const bullet = first ? "- " : "  ";
            first = false;
            if (Array.isArray(itemValue)) {
              if (!itemValue.length) {
                lines.push(`${prefix}  ${bullet}${itemKey}: []`);
              } else {
                lines.push(`${prefix}  ${bullet}${itemKey}:`);
                for (const nested of itemValue) {
                  if (nested && typeof nested === "object" && !Array.isArray(nested)) {
                    let nestedFirst = true;
                    for (const nestedKey of SOURCE_KEY_ORDER) {
                      if (!(nestedKey in nested)) continue;
                      const nestedBullet = nestedFirst ? "- " : "  ";
                      nestedFirst = false;
                      const nestedValue = nested[nestedKey];
                      if (nestedValue && typeof nestedValue === "object" && !Array.isArray(nestedValue)) {
                        lines.push(`${prefix}      ${nestedBullet}${nestedKey}:`);
                        const subOrder = nestedKey === "evidence_range"
                          ? EVIDENCE_RANGE_KEY_ORDER
                          : Object.keys(nestedValue);
                        for (const subKey of subOrder) {
                          if (!(subKey in nestedValue)) continue;
                          lines.push(`${prefix}          ${subKey}: ${quoteScalar(nestedValue[subKey])}`);
                        }
                      } else {
                        lines.push(
                          `${prefix}      ${nestedBullet}${nestedKey}: ${quoteScalar(nestedValue)}`,
                        );
                      }
                    }
                  } else {
                    lines.push(`${prefix}      - ${quoteScalar(nested)}`);
                  }
                }
              }
            } else {
              lines.push(`${prefix}  ${bullet}${itemKey}: ${quoteScalar(itemValue)}`);
            }
          }
        } else {
          lines.push(`${prefix}  - ${quoteScalar(item)}`);
        }
      }
    } else {
      lines.push(`${prefix}${key}: ${quoteScalar(value)}`);
    }
  }
}

function emitYaml(provenance) {
  if (!provenance || typeof provenance !== "object" || Array.isArray(provenance)) {
    throw new YamlCodecError("provenance must be an object");
  }
  const lines = ["---", "docforge_provenance:"];
  emitMapping(lines, provenance, PROVENANCE_KEY_ORDER, 2);
  lines.push("---", "");
  return lines.join("\n");
}

function emitDocumentFrontmatter(docId, title, provenance, description = null) {
  if (!docId || !title) throw new YamlCodecError("document id and title must be non-empty");
  const lines = ["---", `id: ${quoteScalar(docId)}`, `title: ${quoteScalar(title)}`];
  if (description !== null && description !== undefined) {
    lines.push(`description: ${quoteScalar(description)}`);
  }
  lines.push("docforge_provenance:");
  emitMapping(lines, provenance, PROVENANCE_KEY_ORDER, 2);
  lines.push("---", "");
  return lines.join("\n");
}

function wrapDocument(provenance, body) {
  let next = body;
  if (next.startsWith("\n")) next = next.replace(/^\n+/, "");
  if (next && !next.endsWith("\n")) next += "\n";
  return emitYaml(provenance) + next;
}

function rejectUnsupported(raw) {
  for (const line of raw.split(/\r?\n/)) {
    const stripped = line.trim();
    if (!stripped || stripped.startsWith("#")) continue;
    if (stripped === "---" || stripped.startsWith("...")) {
      throw new YamlCodecError("unsupported yaml construct: multi-document marker");
    }
    if (stripped.includes("&") || stripped.includes("*")) {
      throw new YamlCodecError("unsupported yaml construct: anchor or alias");
    }
    if (
      stripped.endsWith("|") || stripped.endsWith(">")
      || stripped.endsWith("|-") || stripped.endsWith(">-")
    ) {
      throw new YamlCodecError("unsupported yaml construct: block scalar");
    }
    if (/:\s*[\[{].+[\]}]/.test(stripped)) {
      throw new YamlCodecError("unsupported yaml construct: non-empty flow collection");
    }
  }
}

function parseScalar(raw) {
  const value = raw.trim();
  if (value === "[]") return [];
  if (value === "{}") return {};
  if (value === "null" || value === "~") return null;
  if (value === "true") return true;
  if (value === "false") return false;
  if (value.length >= 2 && value[0] === value[value.length - 1] && (value[0] === '"' || value[0] === "'")) {
    if (value[0] === '"') return JSON.parse(value);
    return value.slice(1, -1);
  }
  return value;
}

function indentOf(line) {
  return line.length - line.trimStart().length;
}

function parseYamlMapping(raw) {
  rejectUnsupported(raw);
  const lines = raw.split(/\r?\n/).filter((line) => line.trim() && !line.trim().startsWith("#"));
  const root = {};
  const stack = [[-1, root]];
  let i = 0;
  while (i < lines.length) {
    const line = lines[i];
    const indent = indentOf(line);
    const content = line.trim();
    while (stack.length > 1 && indent <= stack[stack.length - 1][0]) stack.pop();
    const parent = stack[stack.length - 1][1];
    if (content.startsWith("- ")) {
      if (!Array.isArray(parent)) throw new YamlCodecError("list item without list parent");
      const itemContent = content.slice(2);
      if (itemContent.includes(": ") || itemContent.endsWith(":")) {
        const item = {};
        parent.push(item);
        stack.push([indent, item]);
        if (itemContent.endsWith(":")) {
          const key = itemContent.slice(0, -1).trim();
          if (i + 1 < lines.length && indentOf(lines[i + 1]) > indent) {
            const nxt = lines[i + 1].trim();
            if (nxt.startsWith("- ")) {
              item[key] = [];
              stack.push([indent + 2, item[key]]);
            } else if (nxt === "[]") {
              i += 1;
              item[key] = [];
            } else {
              item[key] = {};
              stack.push([indent + 2, item[key]]);
            }
          } else {
            item[key] = {};
          }
        } else {
          const idx = itemContent.indexOf(":");
          const key = itemContent.slice(0, idx).trim();
          item[key] = parseScalar(itemContent.slice(idx + 1));
        }
      } else {
        parent.push(parseScalar(itemContent));
      }
      i += 1;
      continue;
    }
    if (content.includes(": ") || content.endsWith(":")) {
      if (!parent || typeof parent !== "object" || Array.isArray(parent)) {
        throw new YamlCodecError("mapping entry without mapping parent");
      }
      if (content.endsWith(":")) {
        const key = content.slice(0, -1).trim();
        if (i + 1 < lines.length && indentOf(lines[i + 1]) > indent) {
          const nxt = lines[i + 1].trim();
          if (nxt === "[]") {
            i += 1;
            parent[key] = [];
          } else if (nxt.startsWith("- ")) {
            parent[key] = [];
            stack.push([indent, parent[key]]);
          } else {
            parent[key] = {};
            stack.push([indent, parent[key]]);
          }
        } else {
          parent[key] = {};
        }
      } else {
        const idx = content.indexOf(":");
        const key = content.slice(0, idx).trim();
        parent[key] = parseScalar(content.slice(idx + 1));
      }
      i += 1;
      continue;
    }
    throw new YamlCodecError(`unsupported yaml construct: ${content}`);
  }
  return root;
}

function splitFrontmatter(text) {
  if (!text.startsWith("---\n")) return { raw: null, body: text, end: 0 };
  const end = text.indexOf("\n---\n", 4);
  if (end < 0) return { raw: null, body: text, end: 0 };
  return {
    raw: text.slice(4, end),
    body: text.slice(end + 5),
    end: end + 5,
  };
}

function parseFrontmatter(text) {
  const { raw, end } = splitFrontmatter(text);
  if (raw == null) return { state: "missing", provenance: null, end: 0 };
  const stripped = raw.trimStart();
  if (stripped.startsWith("{")) {
    let data;
    try {
      data = JSON.parse(raw);
    } catch {
      return { state: "unparseable", provenance: null, end };
    }
    const provenance = data && typeof data === "object" ? data.docforge_provenance : null;
    if (!provenance || typeof provenance !== "object" || Array.isArray(provenance)) {
      return { state: "missing", provenance: null, end };
    }
    if (!("schema" in provenance)) return { state: "legacy", provenance, end };
    return { state: "obsolete", provenance, end };
  }
  let data;
  try {
    data = parseYamlMapping(raw);
  } catch {
    return { state: "unparseable", provenance: null, end };
  }
  const provenance = data && typeof data === "object" ? data.docforge_provenance : null;
  if (!provenance || typeof provenance !== "object" || Array.isArray(provenance)) {
    return { state: "missing", provenance: null, end };
  }
  if (!("schema" in provenance)) return { state: "legacy", provenance, end };
  if (!SUPPORTED_SCHEMA_VERSIONS.has(provenance.schema) || "tool_version" in provenance) {
    return { state: "obsolete", provenance, end };
  }
  return { state: "ok", provenance, end };
}

function rewriteFrontmatter(text, provenance) {
  const { raw, body } = splitFrontmatter(text);
  if (raw == null) return wrapDocument(provenance, text);
  return emitYaml(provenance) + body;
}

module.exports = {
  SCHEMA_VERSION,
  SUPPORTED_SCHEMA_VERSIONS,
  LEGACY_SCHEMA,
  GENERATOR_NAME,
  GENERATOR_VERSION,
  PROVENANCE_FIELDS,
  OPTIONAL_FIELDS,
  SOURCE_ROLES,
  FLOW_VALUES,
  BLOB,
  SCAFFOLD_TOKEN,
  CONTENT_HASH,
  YamlCodecError,
  quoteScalar,
  contentHash,
  scaffoldProvenance,
  inferSourceRole,
  inferProviderFromSnapshot,
  normalizeSections,
  migrateV1ToV2,
  emitYaml,
  emitDocumentFrontmatter,
  wrapDocument,
  parseYamlMapping,
  splitFrontmatter,
  parseFrontmatter,
  rewriteFrontmatter,
};
